src/main/scala/cromwell/engine/WorkflowActor.scala

```diff
@@ -1,79 +1,80 @@
   case object GetOutputs extends WorkflowActorMessage
 
-  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) =
+  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) = {
     Props(new WorkflowActor(id, binding, coercedInputs, backend))
+  }
 
   implicit val ActorTimeout = Timeout(5 seconds)
-}
-
-/**
- * Actor to manage the execution of a single workflow.
- */
-case class WorkflowActor private(id: WorkflowId,
-                                 binding: WdlBinding,
-                                 actualInputs: WorkflowCoercedInputs,
-                                 backend: Backend) extends Actor {
 
-  private val log = Logging(context.system, this)
+  sealed trait WorkflowActorData
+  case object NoData extends WorkflowActorData // Irritating that it takes 2 lines to not use a feature
+}
 
+case class WorkflowActor(id: WorkflowId,
+                         binding: WdlBinding,
+                         actualInputs: WorkflowCoercedInputs,
+                         backend: Backend) extends LoggingFSM[WorkflowState, WorkflowActorData] {
+  implicit val timeout = Timeout(5 seconds)
   private val storeActor = context.actorOf(StoreActor.props(binding, actualInputs))
 
-  def receive: Receive = unstarted
+  startWith(WorkflowSubmitted, NoData)
 
-  /** Unstarted `Receive` state that only responds to `Start` messages. */
-  def unstarted: Receive = LoggingReceive {
-    case WorkflowActor.Start => handleInitialStart()
-    case bad => fail(bad, "unstarted")
+  when(WorkflowSubmitted) {
+    case Event(WorkflowActor.Start, _) => goto(WorkflowRunning)
   }
 
-  /** Started `Receive` state. */
-  def started: Receive = LoggingReceive {
-    case CallActor.Started(call) =>
+  when(WorkflowRunning) {
+    case Event(CallActor.Started(call), _) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Running)
+      stay()
 
-    case CallActor.Completed(completedCall, callOutputs) =>
-      handleCallCompleted(completedCall, callOutputs)
+    case Event(CallActor.Completed(completedCall, callOutputs), _) =>
+      storeActor ! CallCompleted(completedCall, callOutputs)
+      startRunnableCalls()
+      stay()
 
-    case CallActor.Failed(call, failure) =>
+    case Event(CallActor.Failed(call, failure), _) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Failed)
-      context.parent ! WorkflowActor.Failed(failure)
+      // FIXME: Losing the 'failure' here ...
+      goto(WorkflowFailed)
+
+    case Event(WorkflowActor.Complete, _) => goto(WorkflowSucceeded)
+  }
 
-    case WorkflowActor.GetOutputs =>
-      (storeActor ? StoreActor.GetOutputs) pipeTo sender
+  when(WorkflowSucceeded) {
+    case Event(WorkflowActor.GetOutputs, _) =>
+      storeActor.ask(StoreActor.GetOutputs) pipeTo sender
+      stay()
+  }
 
-    case bad => fail(bad, "started")
+  when(WorkflowFailed) { // FIXME: Does this even make sense? If not, remove this
```

Comment:
For my own edification - if WorkflowFailed, what does a GetOutputs mean? Retrieve all outputs which had actually been completed?